src/timeout.ts

```diff
@@ -108,3 +108,131 @@ export class Timeout extends Promise<never> {
     );
   }
 }
+
+/** @internal */
+export type TimeoutContextOptions = {
+  timeoutMS?: number;
+  serverSelectionTimeoutMS: number;
+  waitQueueTimeoutMS: number;
+  socketTimeoutMS?: number;
+};
+
+/** @internal */
+export abstract class TimeoutContext {
+  private _csotEnabled: boolean;
+  clearServerSelectionTimeout: boolean;
+  clearConnectionCheckoutTimeout: boolean;
+
+  constructor(options: TimeoutContextOptions) {
+    this._csotEnabled = options.timeoutMS != null;
+    this.clearServerSelectionTimeout = true;
+    this.clearConnectionCheckoutTimeout = true;
+  }
+
+  static create(options: TimeoutContextOptions): TimeoutContext {
+    if (options.timeoutMS != null) return new CSOTTimeoutContext(options);
+    else return new LegacyTimeoutContext(options);
+  }
+
+  get serverSelectionTimeout(): Timeout | null {
+    return null;
+  }
```

Comment:
If we're going to use an abstract class to define the interface, let's keep all logic on it to a minimum.  It's generally best practice for abstract classes to only have implementations that will be shared across all subclasses (i.e., if there's a method that can be written in terms of the class' abstract members, then you can put it in the abstract class instead of duplicating it in the children).


```suggestion
  abstract get serverSelectionTimeout(): Timeout | null;
  abstract get connectionCheckoutTimeout(): Timeout | null;
  abstract get csotEnabled(): this is CSOTTimeoutContext;
```

Reply:
Don't know how I forgot the abstract keyword doesn't just apply to the class itself 😅